### PR TITLE
fix(next-lint): update option --report-unused-disable-directives to --report-unused-disable-directives-severity

### LINKED ISSUE #64402: 14.2 breaks lint option --report-unused-disable-directives
### Link to the code that reproduces this issue

https://codesandbox.io/p/devbox/stoic-morse-kdqd86

### To Reproduce

1. use next 14.2.0
2. update lint command to be `"lint": "next lint --report-unused-disable-directives error"`
3. npm run lint

### Current vs. Expected behavior

Up until 14.1.4 it lints properly and reports unused directives. 14.2.0 throws an error "Invalid project directory provided, no such directory: /workspace/error". For some reason it appends the "error" to the project dir

### Provide environment information

```bash
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 23.4.0: Fri Mar 15 00:12:49 PDT 2024; root:xnu-10063.101.17~1/RELEASE_ARM64_T6020
  Available memory (MB): 65536
  Available CPU cores: 12
Binaries:
  Node: 20.11.0
  npm: 10.5.0
  Yarn: 1.22.19
  pnpm: 8.15.6
Relevant Packages:
  next: 14.2.0 // Latest available version is detected (14.2.0).
  eslint-config-next: 14.2.0
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.4.5
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

CLI (create-next-app), ESLint (eslint-config-next)

### Which stage(s) are affected? (Select all that apply)

Other (Deployed)

### Additional context

This seems to start happening in 14.2.0-canary.0 and it seems to be connected to updating the commander because the `directory` here https://github.com/vercel/next.js/blob/c9c074c18d8d67f891672f361bc7084d512bec26/packages/next/src/bin/next.ts#L285 now happens to be 'error'

### PR DESCRIPTION
**Pull request by coltonehrman**
## Why?

The Option name and type has been incorrect for `--report-unused-disable-directives` (it's likely that https://github.com/vercel/next.js/pull/61877 exposed this.

For correctness, we have updated the name to `--report-unsed-disable-directives-severity` → https://eslint.org/docs/v8.x/use/command-line-interface#--report-unused-disable-directives-severity.

- Fixes https://github.com/vercel/next.js/issues/64402

### CODE REVIEW

**Comment by ijjk** [comment]
Allow CI Workflow Run

- [ ] approve CI run for commit: 551a9d00ad3a52a0ef20520525cf4d6a2208f6df

Note: this should only be enabled once the PR is ready to go and can only be enabled by a maintainer

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 551a9d00ad3a52a0ef20520525cf4d6a2208f6df

`pnpm test-start test/e2e/middleware-rewrites/test/index.test.ts`
- Middleware Rewrite > should handle catch-all rewrite correctly

<details>
<summary>Expand output</summary>

  ● Middleware Rewrite › should handle catch-all rewrite correctly

    expect(received).toBe(expected) // Object.is equality

    Expected: true
    Received: false

      38 |
      39 |         await retry(async () => {
    > 40 |           expect(requests.some((item) => item.includes('article'))).toBe(true)
         |                                                                     ^
      41 |         })
      42 |       }
      43 |

      at toBe (e2e/middleware-rewrites/test/index.test.ts:40:69)
      at fn (lib/next-test-utils.ts:730:20)
      at Object.<anonymous> (e2e/middleware-rewrites/test/index.test.ts:39:9)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong></summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | coltonehrman/next.js fix/next-cli-lint | Change |
| - | - | - | - |
| buildDuration | 13.9s | 14s | N/A |
| buildDurationCached | 7.5s | 6.1s | N/A |
| nodeModulesSize | 199 MB | 199 MB | N/A |
| nextStartRea..uration (ms) | 396ms | 410ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | coltonehrman/next.js fix/next-cli-lint | Change |
| - | - | - | - |
| 2453-HASH.js gzip | 31.4 kB | 31.4 kB | N/A |
| 3304.HASH.js gzip | 181 B | 181 B | ✓ |
| 3f784ff6-HASH.js gzip | 53.7 kB | 53.7 kB | ✓ |
| 8299-HASH.js gzip | 5.1 kB | 5.1 kB | N/A |
| framework-HASH.js gzip | 45.2 kB | 45.2 kB | ✓ |
| main-app-HASH.js gzip | 242 B | 242 B | ✓ |
| main-HASH.js gzip | 32.2 kB | 32.2 kB | N/A |
| webpack-HASH.js gzip | 1.68 kB | 1.68 kB | N/A |
| Overall change | 99.3 kB | 99.3 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | coltonehrman/next.js fix/next-cli-lint | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | coltonehrman/next.js fix/next-cli-lint | Change |
| - | - | - | - |
| _app-HASH.js gzip | 196 B | 197 B | N/A |
| _error-HASH.js gzip | 184 B | 184 B | ✓ |
| amp-HASH.js gzip | 505 B | 505 B | ✓ |
| css-HASH.js gzip | 324 B | 325 B | N/A |
| dynamic-HASH.js gzip | 2.5 kB | 2.5 kB | N/A |
| edge-ssr-HASH.js gzip | 258 B | 258 B | ✓ |
| head-HASH.js gzip | 352 B | 352 B | ✓ |
| hooks-HASH.js gzip | 370 B | 371 B | N/A |
| image-HASH.js gzip | 4.27 kB | 4.27 kB | ✓ |
| index-HASH.js gzip | 259 B | 259 B | ✓ |
| link-HASH.js gzip | 2.67 kB | 2.67 kB | N/A |
| routerDirect..HASH.js gzip | 314 B | 312 B | N/A |
| script-HASH.js gzip | 386 B | 386 B | ✓ |
| withRouter-HASH.js gzip | 309 B | 309 B | ✓ |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 6.63 kB | 6.63 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | coltonehrman/next.js fix/next-cli-lint | Change |
| - | - | - | - |
| _buildManifest.js gzip | 483 B | 485 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | coltonehrman/next.js fix/next-cli-lint | Change |
| - | - | - | - |
| index.html gzip | 529 B | 528 B | N/A |
| link.html gzip | 541 B | 541 B | ✓ |
| withRouter.html gzip | 524 B | 523 B | N/A |
| Overall change | 541 B | 541 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | coltonehrman/next.js fix/next-cli-lint | Change |
| - | - | - | - |
| edge-ssr.js gzip | 95.6 kB | 95.6 kB | N/A |
| page.js gzip | 3.06 kB | 3.06 kB | ✓ |
| Overall change | 3.06 kB | 3.06 kB | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | coltonehrman/next.js fix/next-cli-lint | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 621 B | 625 B | N/A |
| middleware-r..fest.js gzip | 155 B | 156 B | N/A |
| middleware.js gzip | 25.5 kB | 25.5 kB | N/A |
| edge-runtime..pack.js gzip | 839 B | 839 B | ✓ |
| Overall change | 839 B | 839 B | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | coltonehrman/next.js fix/next-cli-lint | Change |
| - | - | - | - |
| app-page-exp...dev.js gzip | 171 kB | 171 kB | ✓ |
| app-page-exp..prod.js gzip | 97.4 kB | 97.4 kB | ✓ |
| app-page-tur..prod.js gzip | 99.2 kB | 99.2 kB | ✓ |
| app-page-tur..prod.js gzip | 93.4 kB | 93.4 kB | ✓ |
| app-page.run...dev.js gzip | 144 kB | 144 kB | ✓ |
| app-page.run..prod.js gzip | 91.9 kB | 91.9 kB | ✓ |
| app-route-ex...dev.js gzip | 21.4 kB | 21.4 kB | ✓ |
| app-route-ex..prod.js gzip | 15.2 kB | 15.2 kB | ✓ |
| app-route-tu..prod.js gzip | 15.2 kB | 15.2 kB | ✓ |
| app-route-tu..prod.js gzip | 14.9 kB | 14.9 kB | ✓ |
| app-route.ru...dev.js gzip | 21.1 kB | 21.1 kB | ✓ |
| app-route.ru..prod.js gzip | 14.9 kB | 14.9 kB | ✓ |
| pages-api-tu..prod.js gzip | 9.55 kB | 9.55 kB | ✓ |
| pages-api.ru...dev.js gzip | 9.82 kB | 9.82 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.55 kB | 9.55 kB | ✓ |
| pages-turbo...prod.js gzip | 22.5 kB | 22.5 kB | ✓ |
| pages.runtim...dev.js gzip | 23.1 kB | 23.1 kB | ✓ |
| pages.runtim..prod.js gzip | 22.5 kB | 22.5 kB | ✓ |
| server.runti..prod.js gzip | 51.3 kB | 51.3 kB | ✓ |
| Overall change | 948 kB | 948 kB | ✓ |

</details>

<details>
<summary><strong>build cache</strong></summary>

|  | vercel/next.js canary  | coltonehrman/next.js fix/next-cli-lint | Change |
| - | - | - | - |
| 0.pack gzip | 1.58 MB | 1.58 MB | N/A |
| index.pack gzip | 107 kB | 106 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>middleware.js</strong></summary>

Diff too large to display
</details>

</details>

</details>
Commit: 551a9d00ad3a52a0ef20520525cf4d6a2208f6df

**Comment by coltonehrman** [comment]
> @coltonehrman Thanks for this <picture data-single-emoji=":pr:" title=":pr:">:pr:</picture>! Adding a test to make sure we don't regress on this in the future.

No problem! :) I'd be happy to add a test as well I just wasn't sure the best way to go about that. Are there any CLI/lint command test setups I can follow?

File: test/integration/eslint/test/next-lint.test.js [hunk 1]
```diff
@@ -211,6 +203,42 @@ describe('Next Lint', () => {
     )
   })
 
+  test.only('verify options name and type with auto-generated help output', async () => {
```
Comment (eps1lon) [inline test/integration/eslint/test/next-lint.test.js:206]:
```suggestion
  test('verify options name and type with auto-generated help output', async () => {
```

**Comment by coltonehrman** [comment]
> @coltonehrman it should be a Boolean actually :thinkies: → https://eslint.org/docs/latest/use/command-line-interface#--report-unused-disable-directives

I tried it as a boolean and got this error:
```
Invalid Options:
- 'reportUnusedDisableDirectives' must be any of "error", "warn", "off", and null
```

Screenshot 2024-04-12 152420

**Comment by samcx** [comment]
@coltonehrman Interesting—are there <picture data-single-emoji=":docs:" title=":docs:">:docs:</picture> not updated? <picture data-single-emoji=":dig-deep:" title=":dig-deep:">:dig-deep:</picture>

**Comment by coltonehrman** [comment]
> @coltonehrman Interesting—are there :docs: not updated? :dig-deep:

Very likely lol
I was trying to read through this https://github.com/eslint/eslint/issues/15466. Almost seems like they were going back and forth on the design and possibly forgot to update the docs?

**Comment by samcx** [comment]
@coltonehrman Looks like this needs to be re-named to https://eslint.org/docs/latest/use/command-line-interface#--report-unused-disable-directives-severity

**Comment by coltonehrman** [comment]
> @coltonehrman Looks like this needs to be re-named to https://eslint.org/docs/latest/use/command-line-interface#--report-unused-disable-directives-severity

What version of ESLint is Next.js using? Maybe the APIs are different?

**Comment by samcx** [comment]
@coltonehrman It doesn't seem like that was updated → https://eslint.org/docs/v8.x/use/command-line-interface#--report-unused-disable-directives-severity. Next.js is still on v8.

**Comment by coltonehrman** [comment]
> @coltonehrman It doesn't seem like that was updated → https://eslint.org/docs/v8.x/use/command-line-interface#--report-unused-disable-directives-severity. Next.js is still on v8.

The docs don't seem to be in sync with the behavior then. The flag that is suppose to be boolean doesn't appear to support a boolean value when ran and results in an error. However, as you pointed out, it does appear that the `*-severity` flag should be used instead. Should we update Next.js to use the latest version of ESLint and then verify the flags again with the latest behavior?

**Comment by samcx** [comment]
@coltonehrman I'll push up the changes to make this `--report-unused-disable-directives-severity` with the correct choices `['error', null, 'off', 'warn']`. We should update the eslint version in another PR (if there isn't already a PR for that).

File: packages/next/src/bin/next.ts [hunk 2]
```diff
@@ -264,9 +264,11 @@ program
     '--no-inline-config',
     'Prevents comments from changing config or rules.'
   )
-  .option(
-    '--report-unused-disable-directives',
```
Comment (coltonehrman) [inline packages/next/src/bin/next.ts:268]: Are we completely removing this option? Based on the docs it still seems to be a valid option?

File: packages/next/src/cli/next-lint.ts [hunk 3]
```diff
@@ -34,7 +34,7 @@ type NextLintOptions = {
   maxWarnings: number
   outputFile?: string
   quiet?: boolean
-  reportUnusedDisableDirectives: string
+  reportUnusedDisableDirectives: 'error' | 'off' | 'warn'
```
[the remark this answers is not in the document]
Comment (samcx) [reply packages/next/src/cli/next-lint.ts:37]: You're right!

**Comment by coltonehrman** [comment]
@samcx Just out of curiosity, what is the reason for taking over control of the lint command via Next.js vs passing through options to the underlying tool (ie: ESLint)? I feel like this would cause more issues and be harder to maintain.

File: packages/next/src/bin/next.ts (hunk 2 above)
Comment (samcx) [reply packages/next/src/bin/next.ts:268]: @coltonehrman Yes, we're removing this. I'll update the <picture data-single-emoji=":docs:" title=":docs:">:docs:</picture> as well!

**Comment by samcx** [comment]
@coltonehrman That's a good question—I think the idea is we give you good defaults. So when you simply do `next lint`, you don't have to worry about adding your own CLI options.

**Review by samcx — APPROVED** [review]
<picture data-single-emoji=":lgtm:" title=":lgtm:">:lgtm:</picture> <picture data-single-emoji=":spiderman-pointing:" title=":spiderman-pointing:">:spiderman-pointing:</picture>

**Pull request merged**